### DISCUSSION TITLE
Spaces in Object and Field Names

### DISCUSSION THREAD

**Discussion by JeremyVyska** [microsoft/alguidelines#206] [BC Best Practices]
I had someone point out to me recently that with Object and Field Names being so tight, with the limited space of 30 and adding in affixes, he didn't understand why any dev used spaces at all in Object or Field Names, since that introduces the requirement for Double Quotes.

I figured I'd strike up a discussion on it here.

What are your thoughts?  Spaces in Object and Field Names?   Natural and good or a waste of precious 30?

**Comment by NKarolak** [comment]
I prefer to use spaces because names with spaces are easier to read. And they align with Microsoft's naming scheme.

When I exceed the length limitation, then I first try applying abbreviations, and only if that is not sufficient or readable, then I remove spaces.

On the other hand, I avoid double quotes for procedures, just like Microsoft. 
Hence, double quotes indicate object or field names, and that does not disturb me.

**Comment by hermannsson** [comment]
I agree with NKarolak, readability is a big part of me using spaces, and of course referencing Microsoft code through the years. I thought it was kind of weird at first coming from .NET but got used to it pretty quick

However it is an interesting discussion, to the user it doesn't matter at all as they only see the captions

On another note, is there a specific reason field names and especially object names (we use prefixes also so the names can get quite weird) can only use 30 letters?

**Comment by ajkauffmann** [comment]
The 30-character limit is for historical reasons. AL carries a lot of legacy, going back to the early days of Navigator (the predecessor of Navision).

I'm not in favor of using spaces or other non-alphanumeric characters in names. Never missed it in C# or JavaScript. For me, AL code is less readable because of the many double quotes, dashes, dots, and abbreviations in names. And 30 characters is not a lot, so why waste space?😄 We are not used to having non-alphanumeric characters in variable names, are we? And variable names are not even limited to 30 characters. So why should non-alphanumeric characters be used in object and field names while not using them in variable names? 

Personally, I've decided to not follow the common practice in the base app. Instead, I make sure to have good captions. Then it suddenly makes sense to have captions. I mean, why have captions that are a copy of the field name? Knowing that the translation file will fall back on the field name anyway... 

Another reason to get rid of spaces is the automatic translation into valid field names for SQL. Spaces and other non-alphanumeric characters are translated to underscores for SQL. My reaction is then: why not define them already as valid SQL names in the table definition? This practice is even dangerous because you may run into duplicate field names for SQL! Didn't check lately, but it happened in the early days when SQL database became an option for Navision.

Just my 2 ¢ 😎

**Reply by curateideas** [reply]
I like @ajkauffmann 's point comparing AL and other languages. My clinging to spaces is habit from C/AL, but it could be changed.

Is there a code action in VS code for refactoring btw?

**Reply by NKarolak** [reply]
Sure it could be changed. 
However, my inner Monk hates when our naming conventions totally break with the standard code (that for sure won't rename everything).

I do get your point, though, guys.

**Reply by waldo1001** [reply]
I like @NKarolak 's approach and apply that as well.

To use @ajkauffmann 's words "30 characters is not a lot, so why waste space" - but I'd say: "30 characters is not a lot, so why not use it all?" 🤷‍♂️🤪
